src/firewall.c

```diff
@@ -89,29 +89,40 @@ unsigned int FW_MARK_MASK;             /**< @brief Iptables mask: bitwise or of
 char *
 arp_get(const char req_ip[])
 {
-	FILE *proc;
-	char ip[INET6_ADDRSTRLEN];
-	char mac[18];
+	int type = 0;
+	int flags = 0;
+	char ip[16] = {0};
+	char mac[18] = {0};
+	char mask[16] = {0};
+	char line[128] = {0};
+	char dev[128] = {0};
 	char *reply = NULL;
+	FILE *proc_arp = NULL;
 
-	if (!(proc = fopen("/proc/net/arp", "r"))) {
+	if (!(proc_arp = fopen("/proc/net/arp", "r"))) {
 		return NULL;
 	}
 
 	/* Skip first line */
-	while (!feof(proc) && fgetc(proc) != '\n');
+	fgets(line, sizeof(line), proc_arp);
 
 	/* Find ip, copy mac in reply */
-	reply = NULL;
-	while (!feof(proc) && (fscanf(proc, " %15[0-9.] %*s %*s %17[A-Fa-f0-9:] %*s %*s", ip, mac) == 2)) {
-		if (strcmp(ip, req_ip) == 0) {
+	while (fgets(line, sizeof(line), proc_arp))
+	{
+		if (sscanf(line, "%s 0x%x 0x%x %s %s %s\n",
```

Comment:
Yes I see that, but this is clearly not the only place it tries to get the mac address and this could be redundant code - I am thinking this as what you have found would fail in any hardware or version of OpenWrt I have used and I would like to have a better understanding of what is going on here.
Currently I am using Chaos Calmer and Barrier Breaker with standard releases for ar71xx and x86 and in all combinations /proc/net/arp is a read only place holder.
What is in your /proc/net/arp file? Does it relate to the ip address in question?